thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/cache/DefaultTimeSeriesCache.java

```diff
@@ -0,0 +1,194 @@
+ * data from centralized cache as an alternative to directly fetching from the
+ * data source each time.
+ */
+
+public class DefaultTimeSeriesCache implements TimeSeriesCache {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultTimeSeriesCache.class);
+
+  private final MetricConfigManager metricDAO;
+  private final DatasetConfigManager datasetDAO;
+  private final QueryCache queryCache;
+  private CouchbaseCacheDAO cacheDAO;
+  private final ExecutorService executor;
+
+  public DefaultTimeSeriesCache(MetricConfigManager metricDAO, DatasetConfigManager datasetDAO, QueryCache queryCache, CouchbaseCacheDAO cacheDAO, ExecutorService executorService) {
+    this.metricDAO = metricDAO;
+    this.datasetDAO = datasetDAO;
+    this.queryCache = queryCache;
+    this.cacheDAO = cacheDAO;
+    this.executor = executorService;
+  }
+
+  /**
+   * Given a ThirdEyeRequest object, builds the CacheRequest object and attempts to fetch data
+   * from the cache. If the requested slice of data is not in the cache or is only partially in
+   * the cache, will fetch the missing slices and build the complete timeseries.
+   * @param thirdEyeRequest ThirdEyeRequest built from aligned metricSlice request.
+   * @return ThirdEyeResponse, the full response for the request
+   * @throws Exception if fetch from original data source was not successful.
+   */
+  public ThirdEyeResponse fetchTimeSeries(ThirdEyeRequest thirdEyeRequest) throws Exception {
+
+    if (!CacheConfig.useCentralizedCache()) {
+      return this.queryCache.getQueryResult(thirdEyeRequest);
+    }
+
+    ThirdEyeCacheResponse cacheResponse = cacheDAO.tryFetchExistingTimeSeries(ThirdEyeCacheRequest.from(thirdEyeRequest));
+
+    DateTime sliceStart = thirdEyeRequest.getStartTimeInclusive();
+    DateTime sliceEnd = thirdEyeRequest.getEndTimeExclusive();
+
+    if (cacheResponse.isMissingSlice(sliceStart.getMillis(), sliceEnd.getMillis())) {
+      fetchMissingSlices(cacheResponse);
+    }
+
+    return buildThirdEyeResponseFromCacheResponse(cacheResponse);
+  }
+
+  /**
+   * Used if cache had partial or no data for the requested time slice. Checks for which
+   * slices are missing and fetches them, then adds them to the response.
+   * @param cacheResponse cache response object
+   * @throws Exception if fetch to data source throws error.
+   */
+  private void fetchMissingSlices(ThirdEyeCacheResponse cacheResponse) throws Exception {
+
+    ThirdEyeRequest request = cacheResponse.getCacheRequest().getRequest();
+    long metricId = request.getMetricFunctions().get(0).getMetricId();
+    long requestSliceStart = request.getStartTimeInclusive().getMillis();
+    long requestSliceEnd = request.getEndTimeExclusive().getMillis();
+
+    ThirdEyeResponse result;
+    MetricSlice slice;
+
+    if (cacheResponse.hasNoRows()) {
+      result = this.queryCache.getQueryResult(cacheResponse.getCacheRequest().getRequest());
+      insertTimeSeriesIntoCache(result);
```

Comment:
How about we make all `insertTimeSeriesIntoCache()` asynchronous? so that it won't impact the performance even in the case of a cache miss.

Reply:
Previously I was seeing that running it synchronously had slightly better performance, but I will take another look